#PRAKTIKUM 8
# Evaluasi Beberapa Algoritma Klasifikasi

Tujuan :
Setelah mengikuti praktikum ini praktikan
1.   Memahasi Evaluasi beberapa metrik untuk mengukur kinerja klasifikasi
2.   Mahasiswa dapat memahami confisin matrik dan perananya untuk evaluasi klasifikasi
3. Memehami dan mampu menggunakan evaluasi k-fold 








Nama  : Gede Indra Adi Brata

NIM   : 225150200111006

Kelas : PPM F

##Dasar Teori

Validasi klasifikasi

Metode klasifikasi dibangun untuk dapat belajar dan diharapakan dapat membuat prediksi atau keputusan suatu data baru berdasarkan data yang sudah ada. Proses pembelajaran terhadap dataset yang sudah ada untuk mendapatkan model. Selanjutnya suatu data baru yang belum teridentifikasi dapat ditentukan label atau kelasnya berdasarkan model yang telah dibangun. 
Proses pembelajaran yang dibangun terhadap suatu datasaet dilakukan dengan cara membagi suatu data menjadi dua bagian yakni sebagian digunakan sebagai data pelatihan dan sebagian lainnya sebagai data validasi. Besaran jumlah masing-masing data pelatihan dan data validasi dapat ditentukan (50%,50%) atau (75%,25%).
Tahap selanjutnya data pelatihan akan digunakan untuk membagun model dengan menentukan parameter terbaik dari masing-masing model berdasrakan dataset yang ada. Kinerja terbaik dari suatu model yang dikembangkan diukur menggunakan beberap tolok ukur sebagai berikut :
- presison
- recall
- specifity
- akurasi
Masing-masing metrik tersebut ditentukan berdasarkan matrik kebingungan.



Confusion matrik
Mtriks kebingungan atau matriks kesalahan adalah suatu tabel yang merepresentasikan kinerja suatu algoritma. Setiap baris matriks mewakili instance di kelas yang diprediksi sementara setiap kolom mewakili instance di kelas yang sebenarnya (atau sebaliknya). 

#Beberapa tahapan dalam valisasi klasifikasi


Beberapa metode klasifikasi yang telah dipelajari mempunyai kelebihan dan kekurannya. Perbandingan kinerja antar metode klasifikasi ini diukur menggunakan beberapa metrik.  Pada praktikum ini akan dibahas beberapa metrik untuk mengukur kinerja klasifikasi. 
Tahapan untuk mengevaluasi suatu metode klasifikasi dilakukan sebagai berikut :
- Pisahkan set data validasi.
- Siapkan test harness untuk menggunakan validasi silang 10 kali lipat.
- Buat beberapa model berbeda untuk memprediksi spesies dari pengukuran bunga
- Pilih model terbaik.

##1. Tentukan metode klasifikasi yang akan dievaluasi

In [1]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.naive_bayes import GaussianNB 
from sklearn.svm import SVC
models = [] 
models.append(('KNN', KNeighborsClassifier())) 
models.append(('NB', GaussianNB())) 
models.append(('SVM', SVC(gamma='auto'))) 

In [2]:
results = [] 
names = []

##2. Baca dataset
Pada contoh ini terdapat tiga buah dataset yakni :
- dataR2
- iris data 
- breast cancer

Kita akan gunakan dataR2 sebagai contoh sedangkan dua data berikunya akan digunakan sabagai latihan.

In [3]:
import pandas as pd 
import numpy as np 

In [4]:
dataset = pd.read_csv('dataR2.csv') 

In [5]:
dataset 

Age        BMI  Glucose  Insulin      HOMA   Leptin  Adiponectin  \
0     48  23.500000       70    2.707  0.467409   8.8071     9.702400   
1     83  20.690495       92    3.115  0.706897   8.8438     5.429285   
2     82  23.124670       91    4.498  1.009651  17.9393    22.432040   
3     68  21.367521       77    3.226  0.612725   9.8827     7.169560   
4     86  21.111111       92    3.549  0.805386   6.6994     4.819240   
..   ...        ...      ...      ...       ...      ...          ...   
111   45  26.850000       92    3.330  0.755688  54.6800    12.100000   
112   62  26.840000      100    4.530  1.117400  12.4500    21.420000   
113   65  32.050000       97    5.730  1.370998  61.4800    22.540000   
114   72  25.590000       82    2.820  0.570392  24.9600    33.750000   
115   86  27.180000      138   19.910  6.777364  90.2800    14.110000   

     Resistin    MCP.1  Classification  
0     7.99585  417.114               1  
1     4.06405  468.786               1  
2     9.27715  554.697               1  
3    12.76600  928.220               1  
4    10.57635  773.920               1  
..        ...      ...             ...  
111  10.96000  268.230               2  
112   7.32000  330.160               2  
113  10.33000  314.050               2  
114   3.27000  392.460               2  
115   4.35000   90.090               2  

[116 rows x 10 columns]

##3.1 Membuat Set Data Validasi
Tujuan evaluasi adalah memperkirakan keakuratan model terbaik pada data tak terlihat dengan mengevaluasinya pada data tak terlihat aktual. Evaluasi ini dilakukan dengan cara membagi dataset menjadi dua bagian yakni sebagian digunakan sebagai data latih dan sebagian yang lain digunakan sebagai data uji. Beberapa praktikum sebelumnya telah dipraktekkan dengan menggunakan metode train_test_split : 
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1)


Sebelum membuat data latih dan data uji. Langkah pertama pisahkan antara data attribute dan data label suatu dataset. Karena datasaet data2R terdiri dari 9 attribut dan 1 label.  Maka kita catat X sebagai data dan Y sebagai label.

In [6]:
from sklearn.model_selection import train_test_split 
array = dataset.to_numpy() 
X = array[:,0:9] 
y = array[:,9] 
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1, shuffle=True)

##3.2 K-Fold Cross Validation
k-Fold Cross-Validation Validasi silang adalah prosedur pengambilan sampel ulang yang digunakan untuk mengevaluasi model pembelajaran mesin pada sampel data terbatas. Prosedur ini memiliki parameter tunggal yang disebut k yang mengacu pada jumlah grup yang akan dipecah menjadi sampel data tertentu. Karena itu, prosedur ini sering disebut k-fold cross-validation. Ketika nilai spesifik untuk k dipilih, nilai tersebut dapat digunakan sebagai pengganti k dalam referensi ke model, seperti k = 10 menjadi validasi silang 10 kali lipat.

Validasi silang terutama digunakan dalam pembelajaran mesin terapan untuk memperkirakan kemampuan model pembelajaran mesin pada data yang tidak terlihat. Yaitu, menggunakan sampel terbatas untuk memperkirakan bagaimana performa model secara umum saat digunakan untuk membuat prediksi pada data yang tidak digunakan selama pelatihan model.

Prosedur umumnya adalah sebagai berikut:
- Acak kumpulan data berdasarkan id
- Pisahkan kumpulan data menjadi kelompok k
- Untuk setiap grup unik:
  - Ambil kelompok sebagai kelompok data uji
  - Ambil kelompok yang tersisa sebagai kumpulan data pelatihan
  - Catat hasil pengujian masing-masing group

K-Fold ini digunakan untuk validasi model yang digunakan. Evaluasi dilakukan pada dataset latih. Pada praktikum ini digunakan k=10, berarti akan membagi dataset latih menjadi 10 bagian. Satu bagian digunakan untuk testing dan sisanya sebagai data latih.

In [7]:
from sklearn.model_selection import cross_val_score 
from sklearn.model_selection import StratifiedKFold 

for name, model in models: 
  kfold = StratifiedKFold(n_splits=10, random_state=1, shuffle=True) 
  cv_results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy') 
  results.append(cv_results) 
  names.append(name) 
  print('%s: %f (%f)' % (name, cv_results.mean(), cv_results.std()))

KNN: 0.487778 (0.143540)
NB: 0.662222 (0.125629)
SVM: 0.465556 (0.047778)


##4. Buat Prediksi

Kita harus memilih algoritma yang akan digunakan untuk membuat prediksi.

##4.1 Membuat Prediksi
Kita dapat menyesuaikan model pada seluruh dataset pelatihan dan membuat prediksi pada dataset validasi.

In [8]:
model = SVC(gamma='auto') 
model.fit(X_train, Y_train) 
predictions = model.predict(X_validation)

# 4.2 Evaluasi Prediksi
Kami dapat mengevaluasi prediksi dengan membandingkannya dengan hasil yang diharapkan dalam set validasi, lalu menghitung akurasi klasifikasi, serta matriks kebingungan dan laporan klasifikasi.

In [9]:
from sklearn.datasets import make_classification 
X, y = make_classification(n_samples=1000, n_features=20, n_informative=15, n_redundant=5, random_state=1) 

In [10]:
from sklearn.metrics import classification_report 
print(classification_report(Y_validation, predictions))

              precision    recall  f1-score   support

         1.0       0.25      1.00      0.40         6
         2.0       0.00      0.00      0.00        18

    accuracy                           0.25        24
   macro avg       0.12      0.50      0.20        24
weighted avg       0.06      0.25      0.10        24



c:\Users\indra\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\indra\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\indra\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average,

##Tugas :

1. Lakukan perbandingan klasifikasi antara SVM, Naive Bayes, KNN untuk dataset iris dan breast cancer
2. Pada kedua dataset lakukan proses normaslisai data dan preproses data untuk menengani data yang hilang jika ada.

1. Perbandingan klasifikasi antara SVM, Naive Bayes, KNN

In [11]:
iris = pd.read_csv('iris.csv')
iris.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

In [12]:
iris.isna().sum()

sepal.length    0
sepal.width     0
petal.length    0
petal.width     0
variety         0
dtype: int64

In [13]:
from sklearn.preprocessing import MinMaxScaler
minmax = MinMaxScaler()
iris[['sepal.length', 'sepal.width', 'petal.length', 'petal.width']] = minmax.fit_transform(iris[['sepal.length', 'sepal.width', 'petal.length', 'petal.width']])

In [14]:
iris

sepal.length  sepal.width  petal.length  petal.width    variety
0        0.222222     0.625000      0.067797     0.041667     Setosa
1        0.166667     0.416667      0.067797     0.041667     Setosa
2        0.111111     0.500000      0.050847     0.041667     Setosa
3        0.083333     0.458333      0.084746     0.041667     Setosa
4        0.194444     0.666667      0.067797     0.041667     Setosa
..            ...          ...           ...          ...        ...
145      0.666667     0.416667      0.711864     0.916667  Virginica
146      0.555556     0.208333      0.677966     0.750000  Virginica
147      0.611111     0.416667      0.711864     0.791667  Virginica
148      0.527778     0.583333      0.745763     0.916667  Virginica
149      0.444444     0.416667      0.694915     0.708333  Virginica

[150 rows x 5 columns]

In [15]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.naive_bayes import GaussianNB 
from sklearn.svm import SVC
models = [] 
models.append(('KNN', KNeighborsClassifier())) 
models.append(('NB', GaussianNB())) 
models.append(('SVM', SVC(gamma='auto'))) 

In [16]:
results = [] 
names = []

In [17]:
from sklearn.model_selection import train_test_split 
array = iris.to_numpy() 
X = array[:,0:4] 
y = array[:,4] 
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1, shuffle=True)

In [18]:
from sklearn.model_selection import cross_val_score 
from sklearn.model_selection import StratifiedKFold 

for name, model in models: 
  kfold = StratifiedKFold(n_splits=10, random_state=1, shuffle=True) 
  cv_results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy') 
  results.append(cv_results) 
  names.append(name) 
  print('%s: %f (%f)' % (name, cv_results.mean(), cv_results.std()))

KNN: 0.950000 (0.040825)
NB: 0.950000 (0.055277)
SVM: 0.941667 (0.053359)


Berdasarkan hasil cross validation yang telah dilakukan terhadap data iris yang telah di normalisasi dan preprocessing maka model yang memiliki kinerja cukup baik terhadap data iris, adalah KNN dan Naive Bayes. Namun apabila dibandingkan nilai standar deviasinya maka KNN memiliki nilai standar deviasi lebih kecil dibandingkan dengan NB yang menandakan tingkat variasi dari KNN lebih kecil sehingga kinerja model yang lebih baik adalah KNN.

In [19]:
breast_cancer = pd.read_csv('breast_cancer.csv')
breast_cancer.head()

id  clump_thickness  cell_size  cell_shape  marginal_adhesion  \
0  1000025              5.0        1.0         NaN                1.0   
1  1002945              5.0        4.0         4.0                5.0   
2  1015425              3.0        1.0         1.0                1.0   
3  1016277              6.0        8.0         8.0                1.0   
4  1017023              4.0        1.0         1.0                3.0   

   epithelial_cell_size  bare_nucklei  bland_chromatin  normal_nucleoli  \
0                   NaN           1.0              3.0              1.0   
1                   7.0           NaN              3.0              NaN   
2                   NaN           2.0              3.0              1.0   
3                   3.0           NaN              3.0              7.0   
4                   NaN           1.0              3.0              1.0   

   mitoses diagnosis  
0      1.0    benign  
1      1.0    benign  
2      1.0    benign  
3      1.0    benign  
4      NaN    benign

In [20]:
breast_cancer.isna().sum()

id                       0
clump_thickness         10
cell_size               10
cell_shape              10
marginal_adhesion       10
epithelial_cell_size    10
bare_nucklei            10
bland_chromatin         10
normal_nucleoli         10
mitoses                 10
diagnosis                0
dtype: int64

In [21]:
breast_cancer.drop(['id'], inplace=True, axis=1)

In [22]:
breast_cancer.dropna(inplace=True)

In [23]:
breast_cancer.isna().sum()

clump_thickness         0
cell_size               0
cell_shape              0
marginal_adhesion       0
epithelial_cell_size    0
bare_nucklei            0
bland_chromatin         0
normal_nucleoli         0
mitoses                 0
diagnosis               0
dtype: int64

In [24]:
feature_columns = breast_cancer.columns[:9]
feature_columns

Index(['clump_thickness', 'cell_size', 'cell_shape', 'marginal_adhesion',
       'epithelial_cell_size', 'bare_nucklei', 'bland_chromatin',
       'normal_nucleoli', 'mitoses'],
      dtype='object')

In [25]:
from sklearn.preprocessing import MinMaxScaler
minmax = MinMaxScaler()
breast_cancer[feature_columns] = minmax.fit_transform(breast_cancer[feature_columns])

In [26]:
breast_cancer

clump_thickness  cell_size  cell_shape  marginal_adhesion  \
10         0.000000   0.000000    0.000000           0.000000   
12         0.444444   0.222222    0.222222           0.222222   
15         0.666667   0.333333    0.555556           0.333333   
17         0.333333   0.000000    0.000000           0.000000   
19         0.555556   0.000000    0.000000           0.000000   
24         0.000000   0.000000    0.000000           0.000000   
26         0.222222   0.111111    0.000000           0.000000   
27         0.444444   0.000000    0.000000           0.000000   
30         0.222222   0.000000    0.000000           0.000000   
39         0.111111   0.444444    0.222222           0.222222   
48         0.333333   0.000000    0.000000           0.222222   
49         0.666667   0.777778    0.666667           0.111111   
52         1.000000   0.222222    0.555556           0.111111   
53         0.444444   0.444444    0.444444           0.777778   
54         1.000000   0.444444    0.444444           0.555556   
58         0.444444   0.111111    0.222222           0.000000   
59         0.888889   0.444444    0.444444           0.111111   
62         0.888889   1.000000    1.000000           0.000000   
63         0.555556   0.222222    0.333333           0.000000   
64         0.000000   0.000000    0.000000           0.000000   
65         1.000000   0.333333    0.111111           0.000000   
67         0.444444   0.222222    0.333333           0.000000   
68         0.777778   0.222222    0.777778           0.222222   
69         0.000000   0.000000    0.000000           0.000000   
70         0.444444   0.000000    0.222222           0.000000   
78         0.222222   0.000000    0.000000           0.000000   
86         0.222222   0.222222    0.555556           0.333333   
90         0.000000   0.000000    0.000000           0.000000   
92         0.888889   0.555556    0.888889           0.111111   
93         0.666667   0.444444    0.555556           1.000000   
95         0.111111   0.222222    0.333333           0.333333   
97         1.000000   1.000000    1.000000           1.000000   
98         0.666667   0.222222    0.333333           0.333333   

    epithelial_cell_size  bare_nucklei  bland_chromatin  normal_nucleoli  \
10              0.000000      0.000000         0.166667         0.000000   
12              0.111111      0.222222         0.333333         0.333333   
15              0.555556      0.000000         0.333333         0.222222   
17              0.111111      0.000000         0.166667         0.000000   
19              0.111111      0.000000         0.166667         0.000000   
24              0.111111      0.000000         0.166667         0.000000   
26              0.000000      0.000000         0.000000         0.000000   
27              0.111111      0.000000         0.000000         0.000000   
30              0.000000      0.000000         0.000000         0.000000   
39              0.555556      0.666667         0.833333         0.444444   
48              0.111111      0.000000         0.166667         0.000000   
49              0.333333      0.777778         0.166667         0.777778   
52              0.222222      0.444444         0.333333         1.000000   
53              1.000000      0.777778         0.833333         0.222222   
54              0.777778      0.777778         0.833333         0.000000   
58              0.555556      1.000000         0.500000         0.000000   
59              0.111111      0.111111         0.500000         0.000000   
62              1.000000      0.777778         0.166667         0.222222   
63              0.444444      0.111111         0.166667         0.888889   
64              0.111111      0.000000         0.000000         0.000000   
65              0.222222      0.111111         0.333333         0.222222   
67              0.777778      1.000000         0.333333         0.888889   
68              0.333333      0.888889       

In [27]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.naive_bayes import GaussianNB 
from sklearn.svm import SVC
models = [] 
models.append(('KNN', KNeighborsClassifier())) 
models.append(('NB', GaussianNB())) 
models.append(('SVM', SVC(gamma='auto'))) 

In [28]:
results = [] 
names = []

In [29]:
from sklearn.model_selection import train_test_split 
array = breast_cancer.to_numpy() 
X = array[:,0:9] 
y = array[:,9] 
X_train, X_validation, Y_train, Y_validation = train_test_split(X, y, test_size=0.20, random_state=1, shuffle=True)

In [30]:
from sklearn.model_selection import cross_val_score 
from sklearn.model_selection import StratifiedKFold 

for name, model in models: 
  kfold = StratifiedKFold(n_splits=10, random_state=1, shuffle=True) 
  cv_results = cross_val_score(model, X_train, Y_train, cv=kfold, scoring='accuracy') 
  results.append(cv_results) 
  names.append(name) 
  print('%s: %f (%f)' % (name, cv_results.mean(), cv_results.std()))

KNN: 0.933333 (0.133333)
NB: 0.816667 (0.189297)
SVM: 0.900000 (0.152753)


Berdasarkan hasil perbandingan model KNN, NB, dan SVM terhadap data breast_cancer yang telah dilakukan proses normalisasi serta preprocessing maka didapatkan model KNN memiliki kinerja lebih baik dibandingkan dengan model yang lain. Hal ini karena model KNN mendapatkan nilai rata-rata cross validation sebesar 0.93